### PR TITLE
[FLINK-18821][network] Report PartitionRequest failures to subsequent…

### PR DESCRIPTION
**Pull request by rkhachatryan**
## What is the purpose of the change

Currently, an exception in `PartitionRequestClientFactory` is propagated to the client directly. 
The future in the internal `PartitionRequestClientFactory` map is not completed exceptionally.
This causes subsequent calls to wait indefinitely for the future to complete (see FLINK-18821).

This PR fixes it by completing the future exceptionally in case of error.

## Verifying this change

Added unit test: `PartitionRequestClientFactoryTest.testFailureReportedToSubsequentRequests`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

**Comment by wsry** [comment]
The fix itself LGTM, but I am afraid there is another issue we may also need to fix. It seems to me the exceptionally completed future in the ```clients``` map is not cleaned up correctly. Am I right? If so, I think it is better to also fix it.

**Comment by rkhachatryan** [comment]
Yes, you are right, exceptionally completed future stays in the map.
But I think this was the original intention in `PartitionRequestClientFactory`. Otherwise, new requests would retry again almost immediately.

I think ideally such a future should be removed after some delay (regardless of client requests). But this seems to be out of scope of this PR (which removes indefinite wait). WDYT about addressing it separately?

**Comment by wsry** [comment]
Addressing the issue separately is also reasonable to me.

File: flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureUtils.java [hunk 1]
```diff
@@ -90,6 +90,21 @@
...
 	}
 
+	/**
+	 * Fakes asynchronous execution by immediately executing the operation and completing the supplied future
+	 * either noramlly or exceptionally.
+	 *
+	 * @param operation to executed
+	 * @param <T> type of the result
+	 */
+	public static <T> void completeFromCallable(CompletableFuture<T> future, Callable<T> operation) {
```
Comment (AHeise) [inline flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureUtils.java:100]: I added a similar method to `FutureUtils` called `runSync` (but it creates a new future). Would it make sense to keep the names in sync? (I have no hard feeling about renaming `runSync`).

Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureUtils.java:100]: `runSync` should probably also use your method.

Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureUtils.java:100]: I checked out `FutureUtils.runSyn` but I couldn't find any usages of it (now and in the original commit too).
So I think we should just remove it, but also in a separate PR.
WDYT @AHeise?

Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureUtils.java:100]: I'd probably go by hotfix commit, doesn't deserve a separate PR.

Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureUtils.java:85]: Wanted to avoid extra CI run :)
OK, added this commit.

**Pull request merged**